Add `ember-modifier` dependency (#1)
SergeAstapov/rfcs: text/0000-element-modifier.md

---
Stage: Accepted
Start Date: 
Release Date: Unreleased
Release Versions:
  ember-source: vX.Y.Z
  ember-data: vX.Y.Z
Relevant Team(s): Ember CLI, Learning
RFC PR: 
---

<!--- 
Directions for above: 

Stage: Leave as is
Start Date: Fill in with today's date, YYYY-MM-DD
Release Date: Leave as is
Release Versions: Leave as is
Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
RFC PR: Fill this in with the URL for the Proposal RFC PR
-->

# Element Modifiers

## Summary

This RFC introduces the concept of user defined element modifiers and proposes
adding [ember-modifier](https://github.com/ember-modifier/ember-modifier)
to the blueprints that back `ember new` and `ember addon`, providing an
officially-supported path for using modifiers out of the box.

This RFC supersedes the original [RFC #353 "Modifiers"][rfc-0353].

## Motivation

Glimmer Components (introduced as part of Ember Octane Edition) as a replacement for Classic Components
are simpler, more ergonomic, and more declarative. In contrary with Classic Components,
Glimmer Components don't have any element/DOM based properties or hooks giving access to DOM.

> This was an intentional move as component baking class gets disconnected from DOM manipulation.

Reviewer: ```suggestion
Ember Octane introduced Glimmer Components as a replacement for Classic
Components. They are simpler, more ergonomic, and more declarative. In contrast
with Classic Components, Glimmer Components don't have any element/DOM based
properties or hooks giving access to DOM. This was an intentional move as
component baking class gets disconnected from DOM manipulation.
```

> Modifiers are similar to template helpers, they are functions or classes that can be used

Reviewer: ```suggestion
Modifiers are similar to template helpers: they are functions or classes that can be used
```

in templates directly using `{{double-curlies}}` syntax. The major difference with modifiers
is that they are applied directly to *elements*:

```handlebars
<button {{effect "fade-in"}}>Save</button>
```

This RFC builds on top of low-level primitives for defining element modifiers
introduced in [RFC #373 "Element Modifier Manager"][rfc-0373].
The `ember-modifier` addon was built on top of those low-level primitives and
provides a convenient API for authoring element modifiers in Ember.

> Element Modifiers introduced as part of Ember Octane edition and first class citizens

Reviewer: ```suggestion
Element Modifiers were introduced as part of Ember Octane edition and are first class citizens
```

in Ember application, like components or helpers. However today, developers need to install
an additional library to be able to define custom modifiers.

> This RFC seeks to fill the gap in Ember.js development mental model by

Reviewer: ```suggestion
This RFC seeks to fill this gap in Ember.js' development mental model by
```

A basic element modifier is defined with the type of `modifier`.
For example these paths would be global element modifiers in an application:

* `app/modifiers/draggable.js`
* `app/modifiers/effect.js`

Similar to helpers, modifiers can be function-based:

```js
// /app/modifiers/on.js
import { modifier } from 'ember-modifier';

export default modifier((element, [eventName, handler]) => {
  element.addEventListener(eventName, handler);

  return () => {
    element.removeEventListener(eventName, handler);
  }
});
```

or class-based

```ts
// /app/modifiers/on.js
import Modifier from 'ember-modifier';
import { registerDestructor } from '@ember/destroyable';

function cleanup(instance: OnModifier) {
  let { element, event, handler } = instance;

  if (element && event && handler) {
    element.removeEventListener(event, handler);

    instance.element = null;
    instance.event = null;
    instance.handler = null;
  }
}

export default class OnModifier extends Modifier {
  element = null;
  event = null;
  handler = null;

  modify(element, [event, handler]) {
    this.addEventListener(element, event, handler);
    registerDestructor(this, this.removeEventListener)

>   }

Reviewer: Let's improve this slightly:

```suggestion
  // Manage teardown
  constructor() {
    super(...arguments);
    registerDestructor(this, cleanup);
  }

  // Run on installation and all changes to tracked state.
  modify(element, [event, handler]) {
    // Clear any previous state.
    cleanup(this);
    
    // Set up the new handling.
    this.addEventListener(element, event, handler);
    registerDestructor(this, this.removeEventListener)
  }
```

  // methods for reuse

>   addEventListener = (element, event, handler) => {

Reviewer: No need for this to be an arrow function.

```suggestion
  addEventListener(element, event, handler) {
```

    // Store the current element, event, and handler for when we need to remove
    // them during cleanup.
    this.element = element;
    this.event = event;
    this.handler = handler;

    element.addEventListener(event, handler);
  };
}
```

While this is slightly more complicated than the function-based version,

> but that complexity comes along with much more *control*.

Reviewer: ```suggestion
that complexity comes along with much more *control*. It also allows us to hook
into Ember's dependency injection system, for example to use services.
```

[RFC #432 "Contextual Helpers and Modifiers"][rfc-0432]

> allows creating one-off modifiers and passing by reference:

Reviewer: ```suggestion
[RFC #432: Contextual Helpers and Modifiers][rfc-0432] allows optional
invocation and partial application of modifiers, and allows modifiers to be
yielded, passed as arguments, etc. [RFC #779: First-Class Component
Templates][rfc-0779] allows defining them locally or importing them directly to
be used in a `<template>` context. This combination makes many patterns much
easier to implement:
```

```js
// /app/components/slide-up-card.js
import Component from '@glimmer/component';
import { modifier } from 'ember-modifier';

const lockBodyScroll = modifier((element, [shouldLockBodyScroll]) => {
  document.body.classList.add('lock-scroll');
  return () => document.body.classList.remove('lock-scroll');
});

export default class SlideUpCard extends Component {
  ['lock-body-scroll'] = lockBodyScroll;

> 

Reviewer: We actually don't need this if we're assuming `<template>`!

```suggestion
```

  @action closeCard (event) {
    event.preventDefault();
    this.args.close();
  }

  <template>
    <div
      class="slide-up-card"
      ...attributes
      {{this.lock-body-scroll}}
    >
      {{yield}}

>     </div>

Reviewer: To show the features together, I suggest a slight tweak here:

```suggestion
    {{#if has-block}}
      {{yield (hash lock=lockBodyScroll onClick=this.closeCard)}}
    {{else}}
      <div
        class="slide-up-card"
        ...attributes
        {{lockBodyScroll}}
      >
        <p>{{@bodyContent}}</p>

        <button {{on "click" this.closeCard}}>Dismiss</button>
      </div>
    {{/if}}
```

   updates, and teardown can be more convenient when using a class.
   (Note that here we should *only* teach the [newly-redesigned API][v3.2.0], not
   the deprecated previous APIs. This will simplify our teaching significantly,
   since we can describe how it matches the `Helper` APIs.)

## Drawbacks

- If we merge [RFC #757 "Default Modifier Manager][rfc-0757],
  it may seem redundant with this RFC.
- There is the potential for confusion between the `ember-modifier` package and
  the `@ember/modifier` package which ships as part of `ember-source`.

## Alternatives

- Introduce [RFC #757 "Default Modifier Manager][rfc-0757]
  creating modifiers and ask developers to install `ember-modifier`
  for more complex use cases.
- Integrate `ember-modifier` into the `@ember/modifier` package directly.
- Introduce a new `@ember/*` or `@glimmer/*` package for the contents of
  `ember-modifier`.

## Unresolved questions

None.

[rfc-0353]: https://github.com/emberjs/rfcs/pull/353
[rfc-0373]: https://emberjs.github.io/rfcs/0373-Element-Modifier-Managers.html
[rfc-0432]: https://emberjs.github.io/rfcs/0432-contextual-helpers.html

> [rfc-0757]: https://github.com/emberjs/rfcs/pull/757

Reviewer: ```suggestion
[rfc-0757]: https://github.com/emberjs/rfcs/pull/757
[rfc-0779]: https://github.com/emberjs/rfcs/pull/779
```
